iptables/lock.go

```diff
@@ -0,0 +1,214 @@
+
+	if l.referenceCount == 0 {
+		// The lock isn't currently held.  Acquire it.
+		lockHandle, err := l.GrabIptablesLocks(
+			l.lockFilePath,
+			"@xtables",
+			l.lockTimeout,
+			l.lockProbeInterval,
+		)
+		if err != nil {
+			// We give the lock plenty of time so err on the side of assuming a
+			// programming bug.
+			log.WithError(err).Panic("Failed to acquire iptables lock")
+		}
+		l.iptablesLockHandle = lockHandle
+	}
+	l.referenceCount++
+}
+
+func (l *SharedLock) Unlock() {
+	l.lock.Lock()
+	defer l.lock.Unlock()
+
+	l.referenceCount--
+	if l.referenceCount < 0 {
+		log.Panic("Unmatched Unlock()")
+	}
+	if l.referenceCount == 0 {
+		log.Debug("Releasing iptables lock.")
+		err := l.iptablesLockHandle.Close()
+		if err != nil {
+			// We haven't done anything with the file or socket so we shouldn't be
+			// able to hit any "deferred flush" type errors from the close.  Panic
+			// since we're not sure what's going on.
+			log.WithError(err).Panic("Error while closing iptables lock.")
+		}
+		l.iptablesLockHandle = nil
+	}
+}
+
+type Locker struct {
+	Lock16 io.Closer
+	Lock14 io.Closer
+}
+
+func (l *Locker) Close() error {
+	var err error
+	if l.Lock16 != nil {
+		err = l.Lock16.Close()
+		if err != nil {
+			log.WithError(err).Error("Error while closing lock file.")
+		}
+	}
+	if l.Lock14 != nil {
+		err14 := l.Lock14.Close()
+		if err14 != nil {
+			log.WithError(err14).Error("Error while closing lock socket.")
+		}
+		if err14 != nil && err == nil {
+			err = err14
+		}
+	}
+	return err
+}
+
+var (
+	Err14LockTimeout = errors.New("Timed out waiting for iptables 1.4 lock")
+	Err16LockTimeout = errors.New("Timed out waiting for iptables 1.6 lock")
+)
+
+func GrabIptablesLocks(lockFilePath, socketName string, timeout, probeInterval time.Duration) (io.Closer, error) {
+	var err error
+	var success bool
+
+	l := &Locker{}
+	defer func(l *Locker) {
+		// Clean up immediately on failure
+		if !success {
+			l.Close()
+		}
+	}(l)
+
+	// Grab both 1.6.x and 1.4.x-style locks; we don't know what the
+	// iptables-restore version is if it doesn't support --wait, so we
+	// can't assume which lock method it'll use.
+
+	// Roughly duplicate iptables 1.6.x xtables_lock() function.
+	f, err := os.OpenFile(lockFilePath, os.O_CREATE, 0600)
+	l.Lock16 = f
+	if err != nil {
+		return nil, fmt.Errorf("failed to open iptables lock %s: %v", lockFilePath, err)
+	}
+
+	startTime := time.Now()
+	for {
+		if err := grabIptablesFileLock(f); err == nil {
+			break
+		}
+		if time.Since(startTime) > timeout {
+			countLockTimeouts.Inc()
+			return nil, Err16LockTimeout
+		}
+		time.Sleep(probeInterval)
+		countLockRetries.Inc()
+	}
+
+	startTime14 := time.Now()
+	for {
+		l.Lock14, err = net.ListenUnix("unix", &net.UnixAddr{Name: socketName, Net: "unix"})
+		if err == nil {
+			break
+		}
+		if time.Since(startTime14) > timeout {
+			countLockTimeouts.Inc()
+			return nil, Err14LockTimeout
+		}
+		time.Sleep(probeInterval)
+		countLockRetries.Inc()
```

Comment:
Feels like it would be useful for the 14 and 16 metrics to be distinct - perhaps with labels.  In a failing system, it's probably going to be one of them in particular that is problematic.